### PR TITLE
Change logging performance improvements

### LINKED ISSUE #2459: Bulk operations do not scale
### Environment
* Nautobot version (Docker tag too if applicable): 1.4.3
* Python version: 3.9
* Database platform, version: All
* Middleware(s): N/A

### Steps to Reproduce
1. Have a large data set with many thousands of devices and/or other objects (>50,000)
2. Attempt to perform a bulk update or bulk delete on a large subset of the inventory (~2000) in the UI. For example, updating a custom field value, changing tags, deleting devices, etc.
3. Wait

### Expected Behavior
The operation should succeed and preferably in a reasonable amount of time and with continual feedback to the user.

### Observed Behavior
The operation is contained within a single HTTP request but this request takes a very long time to process, so the user gets no feedback that Nautobot is actually doing anything. Depending on the size of the operation, the request will eventually time out and the user is presented with an ugly 504. I call out the UI specifically here, but the same fundamental issue exists with the REST API.

### PR DESCRIPTION
**Pull request by gsnider2195**
# Closes #2459
# What's Changed

## Global change logging behavior changes

- A cache of all change logging operations are stored in the thread local change context, to allow us to skip database hits when looking up related changes when none exist

## Deferred change logging for bulk edit

- `nautobot.extras.context_managers.deferred_change_logging_for_bulk_operation` context manager introduced for use with bulk edit
  - Defers change logging to the end of the context manager and uses `bulk_create` to create the `ObjectChange` instances, slightly improving performance

## Deferred change logging for bulk delete

- `nautobot.extras.utils.bulk_delete_with_bulk_change_logging` function introduced to handle bulk change log creation for bulk delete. This couldn't be a context manager because we need to have the instance so we can create the ObjectChange before the delete happens

# Screenshots

Test results for creating 100 namespaces, then updating the description on those namespaces 50 times and averaging the results:

image

image

image

I also prototyped deferred change logging for bulk delete but it needs a different code path (around 30 LOC):

image

image

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/extras/utils.py [hunk 1]
```diff
@@ -610,3 +612,35 @@ def migrate_role_data(
...
+            for obj in qs.iterator():
+                if not hasattr(obj, "to_objectchange"):
+                    break
+                oc = obj.to_objectchange(ObjectChangeActionChoices.ACTION_DELETE)
+                oc.user = change_context.user
+                oc.request_id = change_context.change_id
+                oc.change_context = change_context.context
+                oc.change_context_detail = change_context.context_detail[:CHANGELOG_MAX_CHANGE_CONTEXT_DETAIL]
+                queued_object_changes.append(oc)
+            ObjectChange.objects.bulk_create(queued_object_changes)
```
Comment (gsnider2195) [inline nautobot/extras/utils.py:642]: These should be created in batches, like the bulk edit

**Review by HanlinMiao — APPROVED** [review]

**Review by HanlinMiao — APPROVED** [review]

File: changes/2459.changed [hunk 2]
```diff
@@ -0,0 +1 @@
+Improved performance of bulk-edit operations by refactoring change logging logic.
```
Comment (glennmatthews) [inline changes/2459.changed:1]:
```suggestion
Improved performance of bulk-edit and bulk-delete UI operations by refactoring change logging logic.
```

File: nautobot/extras/utils.py [hunk 3]
```diff
@@ -610,3 +612,38 @@ def migrate_role_data(
...
+                    ObjectChange.objects.bulk_create(queued_object_changes)
+                    queued_object_changes = []
+                oc = obj.to_objectchange(ObjectChangeActionChoices.ACTION_DELETE)
+                oc.user = change_context.user
+                oc.request_id = change_context.change_id
+                oc.change_context = change_context.context
+                oc.change_context_detail = change_context.context_detail[:CHANGELOG_MAX_CHANGE_CONTEXT_DETAIL]
+                queued_object_changes.append(oc)
+            ObjectChange.objects.bulk_create(queued_object_changes)
+            return qs.delete()
```
Comment (glennmatthews) [inline nautobot/extras/utils.py:646]: I'll bet you can get better performance here by temporarily unhooking the `_handle_deleted_object` signal altogether, similar to what was done in #5442?

Comment (gsnider2195) [reply nautobot/extras/utils.py:646]: We would need to manually delete ContactAssociations and Notes and increment metrics here if we disconnect the signal. Is the delete signal making Django do extra checks on delete?

Comment (glennmatthews) [reply nautobot/extras/utils.py:646]: Good points, out of scope for this PR then. But the short answer is yes, it does.

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**